go/ct/gen/interval_solver_test.go

```diff
@@ -0,0 +1,326 @@
+			setup: func(s *IntervalSolver[int32]) {
+				s.Exclude(350, 450)
+			},
+			want: "X ∈ [200..349]",
+		},
+		"fragment-range": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.Exclude(250, 300)
+				s.Exclude(320, 380)
+			},
+			want: "X ∈ [200..249] ∪ [301..319] ∪ [381..400]",
+		},
+		"fragment-range-with-overlap": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.Exclude(250, 300)
+				s.Exclude(320, 380)
+				s.Exclude(220, 310)
+			},
+			want: "X ∈ [200..219] ∪ [311..319] ∪ [381..400]",
+		},
+		"remove-empty-interval": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.Exclude(300, 250) // < considered empty
+			},
+			want: "X ∈ [200..400]",
+		},
+		"equality-constraint": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.AddEqualityConstraint(300)
+			},
+			want: "X ∈ [300..300]",
+		},
+		"lower-boundary-constraint": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.AddLowerBoundary(300)
+			},
+			want: "X ∈ [300..400]",
+		},
+		"upper-boundary-constraint": {
+			setup: func(s *IntervalSolver[int32]) {
+				s.AddUpperBoundary(300)
+			},
+			want: "X ∈ [200..300]",
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			solver := NewIntervalSolver[int32](200, 400)
+			test.setup(solver)
+			got := solver.String()
+			if got != test.want {
+				t.Errorf("unexpected result, got %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+func TestIntervalSolver_Exclude(t *testing.T) {
+	const N = 20
+	// remove interval [c..d] from [a..b]
+	for a := 0; a < N; a++ {
+		for b := 0; b < N; b++ {
+			for c := 0; c < N; c++ {
+				for d := 0; d < N; d++ {
+					solver := NewIntervalSolver(uint32(a), uint32(b))
+					solver.Exclude(uint32(c), uint32(d))
+					for i := 0; i < N; i++ {
+						inAtoB := a <= i && i <= b
+						inCtoD := c <= i && i <= d
+						want := inAtoB && !inCtoD
+						got := solver.Contains(uint32(i))
+						if got != want {
+							t.Fatalf("unexpected result for %d ∈ [%d..%d] \\ [%d..%d]: got %v, want %v", i, a, b, c, d, got, want)
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func TestIntervalSolver_EmptyInterval(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 100)
+	_, err := solver.Generate(rand.New())
+	if err != ErrUnsatisfiable {
+		t.Errorf("failed to generate point in empty interval: %v", err)
+	}
+}
+
+func TestIntervalSolver_ExcludeEmptyIntervalHasNoEffect(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	clone := solver.Clone()
+	clone.Exclude(300, 250)
+	if !solver.Equals(clone) {
+		t.Fatalf("empty interval should not be modified by adding empty interval")
+	}
+}
+
+func TestIntervalSolver_TestIntervals(t *testing.T) {
+	solver := NewIntervalSolver[int32](200, 400)
+	solver.Exclude(250, 300)
+
+	rnd := rand.New()
+	for i := 0; i < 100; i++ {
+		res, err := solver.Generate(rnd)
+		if err != nil {
+			t.Fatalf("error solving intervals: %v", err)
+		}
+		if res < 200 || res > 400 || (res >= 250 && res <= 300) {
```

Comment:
There should be a test case making sure that the boundaries are hit by the generation process.